content/posts/flp-inside-out.md

```diff
@@ -0,0 +1,78 @@
+
+//TODO: include a tesseract of 
+
+#### Internal state of the processes are not important
+
+But what about internal state of the process? Why isn't it mentioned in anywhere in the lemma? Does this imply that each process must start with the same internal state? Can we "break" the proof by coming up with an alogrithm that starts off with processes having different internal state? 
+
+The answer is ... we don't care if all processes have the same internal state or not. If an alogrithm is deterministic it has to define the internal state of each process before any of the values for `x` registers get defined. Futhermore, the state can depend only on some knowledge available before the algorithm starts e.g. process identifier. What this means is that even if each process has different internal state, these (different) states will be same for each possible permutation of the `x` register values if the algorithm is deterministic. 
+
+
+### Lemma 3
+
+#### C0 and C1 that are one step apart
+
+> (...) D contains both 0-valent and 1-valent configurations.
+> Call two configurations neighbors if one results from the other in a single step. By an easy induction, there exist neighbors Co, Ci E, such that Di = e(Ci) is i-valent, i = 0, 1.
+
+Why is it true that there have to be C0 and C1 that are only one step aparts? Well if C is the configuration we start from than all reachable configurations through schedules without `e` form a DAG. Nodes are configurations of the system and edges connect configurations if there is an event that can lead from configuraiton to the other. D as defined in the paper can be "visialized" as a layer of nodes atop the DAG (reachable via `e` event). We already know that D contains both 0-valent and 1-valent configuration (let's call them D0' and D1'). Let's consider a path between C0' and C1' nodes that connect to D0' and D1' via `e`. There must be an edge `e'` on that path that connects nodes with neighroubs in D that have different valence.
+
+//TODO: a layered DAG visualization
+
+### Proof
+
+#### Message ordering inside Message Buffer
+
+> Suppose then that configuration C is bivalent and that process p heads the priority queue. Let m be the earliest message to p in C’s message buffer, if any, and 0 otherwise.
+
+The final proof depends on ordering inside Message Buffer. It is assumed that Message Buffer orders messages based on the send time. Why is that important, would it be possible to skip that part and e.g. prove that for any bi-variant configuration C there is some schedule sigma such that sigma(C) is bi-variant? 
+
+The answer is .. no. The ordering in Message Buffer and the fact that each in each round we take the oldest message for a given process enusers that each message is eventually delivered. If we can't show that each message is eventually delivered we can't be sure that a given schedule is admissable.
```

Comment:
I feel like this is a trick. Why don't we just assume that each message is eventually delivered and drop the requirement that `receive` must take the oldest one? So it could take any message but eventually it'd need to take all of them.